agent/service_manager.go

```diff
@@ -0,0 +1,237 @@
+package agent
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/consul/agent/cache"
+	cachetype "github.com/hashicorp/consul/agent/cache-types"
+	"github.com/hashicorp/consul/agent/structs"
+	"golang.org/x/net/context"
+)
+
+// The ServiceManager is a layer for service registration in between the agent
+// and the local state. Any services must be registered with the ServiceManager,
+// which then maintains a long-running watch of any globally-set service or proxy
+// configuration that applies to the service in order to register the final, merged
+// service configuration locally in the agent state.
+type ServiceManager struct {
+	services map[string]*serviceConfigWatch
+	agent    *Agent
+
+	sync.Mutex
+}
+
+func NewServiceManager(agent *Agent) *ServiceManager {
+	return &ServiceManager{
+		services: make(map[string]*serviceConfigWatch),
+		agent:    agent,
+	}
+}
+
+// AddService starts a new serviceConfigWatch if the service has not been registered, and
+// updates the existing registration if it has. For a new service, a call will also be made
+// to fetch the merged global defaults that apply to the service in order to compose the
+// initial registration.
+func (s *ServiceManager) AddService(service *structs.NodeService, chkTypes []*structs.CheckType, persist bool, token string, source configSource) error {
+	reg := serviceRegistration{
+		service:  service,
+		chkTypes: chkTypes,
+		persist:  persist,
+		token:    token,
+		source:   source,
+	}
+
+	// If a service watch already exists, update the registration. Otherwise,
+	// start a new config watcher.
+	s.Lock()
+	watch, ok := s.services[service.ID]
+	s.Unlock()
+	if ok {
+		if err := watch.updateRegistration(&reg); err != nil {
+			return err
```

Comment:
What happens here if `updateRegistration` returns a temporary error (like servers disconnected)? It seems like the error will propagate back to the user (good) but the agent will treat an error to mean that it should "clean up" the partial registration and actually remove the service (bad).

I've not checked that out yet - it might be this is the same behaviour as before which is OK but just noting it while I read through.

Reply:
The only errors that can come back from `updateRegistration` are from the call to `addServiceInternal` which only has the code to add the service/checks to the local state, so there shouldn't be a behavior change there.